fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneStoredFieldsTest.java

```diff
@@ -0,0 +1,395 @@
+            if (useOptimizedStoredFieldsFormat) {
+                try (FDBDirectory directory = new FDBDirectory(recordStore.indexSubspace(index), context, index.getOptions())) {
+                    assertDocCountPerSegment(directory, List.of("_0"), List.of(3));
+                }
+                assertTrue(timer.getCounter(LuceneEvents.Waits.WAIT_LUCENE_GET_STORED_FIELDS).getCount() > 1);
+                assertTrue(timer.getCounter(LuceneEvents.Counts.LUCENE_WRITE_STORED_FIELDS).getCount() >= 3);
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @BooleanSource
+    void testInsertDeleteDocuments(boolean useOptimizedStoredFieldsFormat) throws Exception {
+        Index index = getSimpleTextIndex(useOptimizedStoredFieldsFormat);
+
+        try (FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            recordStore.saveRecord(createSimpleDocument(1623L, "Document 1", 2));
+            recordStore.saveRecord(createSimpleDocument(1624L, "Document 2", 2));
+            recordStore.saveRecord(createSimpleDocument(1547L, "NonDocument 3", 2));
+            context.commit();
+        }
+        try (FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            recordStore.deleteRecord(Tuple.from(1623L));
+            context.commit();
+        }
+        try (FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            final RecordQuery query = buildQuery("Document", Collections.emptyList(), SIMPLE_DOC);
+            queryAndAssertFields(query, "text", Set.of(1624L), Set.of("Document 2"));
+        }
+    }
+
+    @ParameterizedTest
+    @BooleanSource
+    void testInsertDeleteDocumentsSameTransaction(boolean useOptimizedStoredFieldsFormat) throws Exception {
+        Index index = getSimpleTextIndex(useOptimizedStoredFieldsFormat);
+
+        try (FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            recordStore.saveRecord(createSimpleDocument(1623L, "Document 1", 2));
+            recordStore.saveRecord(createSimpleDocument(1624L, "Document 2", 2));
+            recordStore.saveRecord(createSimpleDocument(1547L, "NonDocument 3", 2));
+
+            recordStore.deleteRecord(Tuple.from(1623L));
+            context.commit();
+        }
+        try (FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            final RecordQuery query = buildQuery("Document", Collections.emptyList(), SIMPLE_DOC);
+            queryAndAssertFields(query, "text", Set.of(1624L), Set.of("Document 2"));
+        }
+    }
+
+    @ParameterizedTest
+    @BooleanSource
+    void testInsertUpdateDocuments(boolean useOptimizedStoredFieldsFormat) throws Exception {
```

Comment:
It would be good to have a test of actually doing merges to make sure the right stuff ends up in the right place.
I think you should be able to simply use 3 transactions for saving the 3 documents, and then assert about the results, and assert that there is only one segment, and it has 3 documents.

Reply:
Added similar test, though the segments do not merge.